samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -65,46 +73,82 @@ public StreamGraphImpl(ApplicationRunner runner, Config config) {
+        new InputOperatorSpec<>(streamSpec, keySerde, valueSerde,
+            (BiFunction<K, V, M>) msgBuilder, this.getNextOpId()));
     return new MessageStreamImpl<>(this, inputOperators.get(streamSpec));
   }
 
   @Override
-  public <K, V, M> OutputStream<K, V, M> getOutputStream(String streamId,
-      Function<? super M, ? extends K> keyExtractor, Function<? super M, ? extends V> msgExtractor) {
-    if (keyExtractor == null) {
-      throw new IllegalArgumentException("keyExtractor can't be null for an output stream.");
+  public <K, V, M> MessageStream<M> getInputStream(String streamId,
+      BiFunction<? super K, ? super V, ? extends M> msgBuilder) {
+    if (defaultKeySerde instanceof NoOpSerde) {
+      LOGGER.info("Using NoOpSerde as the default key serde for input stream " + streamId +
+          ". Keys will not be deserialized");
     }
-
-    if (msgExtractor == null) {
-      throw new IllegalArgumentException("msgExtractor can't be null for an output stream.");
+    if (defaultMsgSerde instanceof NoOpSerde) {
+      LOGGER.info("Using NoOpSerde as the default msg serde for input stream " + streamId +
+          ". Values will not be deserialized");
     }
+    return getInputStream(streamId, (Serde<K>) defaultKeySerde, (Serde<V>) defaultMsgSerde, msgBuilder);
+  }
 
-    if (outputStreams.containsKey(runner.getStreamSpec(streamId))) {
-      throw new IllegalStateException("getOutputStream() invoked multiple times "
-          + "with the same streamId: " + streamId);
-    }
+  @Override
+  public <K, V, M> OutputStream<K, V, M> getOutputStream(String streamId,
+      Serde<K> keySerde, Serde<V> msgSerde,
+      Function<? super M, ? extends K> keyExtractor, Function<? super M, ? extends V> msgExtractor) {
+    Preconditions.checkNotNull(keySerde, "keySerde must not be null for an output stream.");
+    Preconditions.checkNotNull(msgSerde, "msgSerde must not be null for an output stream.");
+    Preconditions.checkNotNull(keyExtractor, "keyExtractor must not be null for an output stream.");
+    Preconditions.checkNotNull(msgExtractor, "msgExtractor must not be null for an output stream.");
+    Preconditions.checkState(!outputStreams.containsKey(runner.getStreamSpec(streamId)),
+        "getOutputStream must not be called multiple times with the same streamId: " + streamId);
 
     StreamSpec streamSpec = runner.getStreamSpec(streamId);
     outputStreams.put(streamSpec,
-        new OutputStreamImpl<>(streamSpec, (Function<M, K>) keyExtractor, (Function<M, V>) msgExtractor));
+        new OutputStreamImpl<>(streamSpec, keySerde, msgSerde,
+            (Function<M, K>) keyExtractor, (Function<M, V>) msgExtractor));
     return outputStreams.get(streamSpec);
   }
 
   @Override
+  public <K, V, M> OutputStream<K, V, M> getOutputStream(String streamId,
+      Function<? super M, ? extends K> keyExtractor, Function<? super M, ? extends V> msgExtractor) {
+    if (defaultKeySerde instanceof NoOpSerde) {
```

Comment:
These checks are repeated a few times. Refactoring them would help us keep the logic (particularly the log messages) consistent.

Reply:
Related question, are the small variances in log messages intentional?